tests/frontend/test_tdmprogram.py

```diff
@@ -0,0 +1,425 @@
+
+    nXvar = np.var(nX)
+    nPvar = np.var(nP)
+
+    assert np.allclose(nXvar, 4 * np.exp(-2 * sq_r), rtol=5 / np.sqrt(n))
+    assert np.allclose(nPvar, 4 * np.exp(-2 * sq_r), rtol=5 / np.sqrt(n))
+
+
+def test_two_dimensional_cluster_denmark():
+    """
+    Two-dimensional temporal-mode cluster state as demonstrated in https://arxiv.org/pdf/1906.08709
+    """
+    np.random.seed(42)
+    sq_r = 3
+    delay1 = 1  # number of timebins in the short delay line
+    delay2 = 12  # number of timebins in the long delay line
+    n = 400  # number of timebins
+    # Size of cluste is n x delay2
+    # first half of cluster state measured in X, second half in P
+    theta_A = [0] * int(n / 2) + [np.pi / 2] * int(n / 2)  # measurement angles for detector A
+    theta_B = theta_A  # measurement angles for detector B
+
+    # 2D cluster
+    prog = tdmprogram.TDMProgram([1, delay2 + delay1 + 1])
+    with prog.context(theta_A, theta_B, shift="end") as (p, q):
+        ops.Sgate(sq_r, 0) | q[0]
+        ops.Sgate(sq_r, 0) | q[delay2 + delay1 + 1]
+        ops.Rgate(np.pi / 2) | q[delay2 + delay1 + 1]
+        ops.BSgate(np.pi / 4, np.pi) | (q[delay2 + delay1 + 1], q[0])
+        ops.BSgate(np.pi / 4, np.pi) | (q[delay2 + delay1], q[0])
+        ops.BSgate(np.pi / 4, np.pi) | (q[delay1], q[0])
+        ops.MeasureHomodyne(p[1]) | q[0]
+        ops.MeasureHomodyne(p[0]) | q[delay1]
+    eng = sf.Engine("gaussian")
+    result = eng.run(prog)
+    samples = result.all_samples
+
+    xA = result.all_samples[0]
+    xB = result.all_samples[1]
+
+    X_A = xA[: n // 2]  # X samples from detector A
+    P_A = xA[n // 2 :]  # P samples from detector A
+    X_B = xB[: n // 2]  # X samples from detector B
+    P_B = xB[n // 2 :]  # P samples from detector B
+
+    # nullifiers defined in https://arxiv.org/pdf/1906.08709.pdf, Eqs. (1) and (2)
+    N = delay2
+    ntot = len(X_A) - delay2 - 1
+    nX = np.array([X_A[k] + X_B[k] - X_A[k + 1] - X_B[k + 1] - X_A[k + N] + X_B[k + N] - X_A[k + N + 1] + X_B[k + N + 1] for k in range(ntot)])
+    nP = np.array([P_A[k] + P_B[k] + P_A[k + 1] + P_B[k + 1] - P_A[k + N] + P_B[k + N] + P_A[k + N + 1] - P_B[k + N + 1] for k in range(ntot)])
+    nXvar = np.var(nX)
+    nPvar = np.var(nP)
+
+    assert np.allclose(nXvar, 8 * np.exp(-2 * sq_r), rtol=5 / np.sqrt(ntot))
+    assert np.allclose(nPvar, 8 * np.exp(-2 * sq_r), rtol=5 / np.sqrt(ntot))
+
+
+def test_two_dimensional_cluster_tokyo():
+    """
+    Two-dimensional temporal-mode cluster state as demonstrated by Universtiy of Tokyo. See: https://arxiv.org/pdf/1903.03918.pdf
+    """
+
+    start = time.time()
+
+    # temporal delay in timebins for each spatial mode
+    delayA = 0
+    delayB = 1
+    delayC = 5
+    delayD = 0
+
+    # concurrent modes in each spatial mode
+    concurrA = 1 + delayA
+    concurrB = 1 + delayB
+    concurrC = 1 + delayC
+    concurrD = 1 + delayD
+
+    N = [concurrA, concurrB, concurrC, concurrD]
```

Comment:
Btw, even though some of this code looks redundant, in my opinion this is the best way to define and understand how many concurrent modes we have in each spatial mode and, thereby, how `N` is defined.